src/python/pants/option/option_types.py

```diff
@@ -0,0 +1,481 @@
+# Copyright 2021 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import annotations
+
+from enum import Enum
+
+# NB: Avoid using "Optional" and use "| None" instead, as "Option" and "Optional" look similar.
+from typing import TYPE_CHECKING, Any, Dict, Generic, Type, TypeVar, cast, overload
+
+from pants.option import custom_types
+from pants.util.frozendict import FrozenDict
+
+if TYPE_CHECKING:
+    from pants.option.subsystem import Subsystem
+
+SubsystemT = TypeVar("SubsystemT", bound="Subsystem")
+OptionT = TypeVar("OptionT", bound="OptionBase")
+PropType = TypeVar("PropType")
+StrT = TypeVar("StrT", str, "str | None")
+IntT = TypeVar("IntT", int, "int | None")
+FloatT = TypeVar("FloatT", float, "float | None")
+BoolT = TypeVar("BoolT", bool, "bool | None")
+EnumT = TypeVar("EnumT", bound=Enum)
+ValueType = TypeVar("ValueType")
+# NB: We don't provide constraints, as our `XListOption` types act like a set of contraints
+ListMemberType = TypeVar("ListMemberType")
+
+
+# ================================================================
+
+
+class OptionBase(Generic[PropType]):
+    """Descriptor for subsystem options.
+
+    Clients shouldn't use this class directly, instead use one of the concrete classes below.
```

Comment:
Should probably be `_OptionBase` then.

Reply:
It is consumed in one other file: the subsystem `register_options` does a `if isinstance(attr, OptionBase)` therefore I wasn't sure if we should mark "private" :/

Reply:
Hm, I'd still prob use private. My 2 cents on "privacy" with Pants is think of our main consumer: plugin authors. If you should use it in plugin code, make sure it's public. Else, probably better as private.

This is also why we have `pants.engine.internal` -- because of that namespacing, we can get away with using no underscore as it's already obvious it's internal.